lib/Dialect/ESI/cosim/cosim_dpi_server/Server.cpp

```diff
@@ -0,0 +1,162 @@
+//===- Server.cpp - Cosim RPC server ----------------------------*- C++ -*-===//
+//
+// Definitions for the RPC server class. Capnp C++ RPC servers are based on
+// 'libkj' and its asyncrony model plus the capnp C++ API, both of which feel
+// very foreign. In general, both RPC arguments and returns are passed as a C++
+// object. In order to return data, the capnp message must be constructed inside
+// that object.
+//
+// A [capnp encoded message](https://capnproto.org/encoding.html) can have
+// multiple 'segments', which is a pain to deal with. (See comments below.)
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/cosim/Server.h"
+#include <capnp/ez-rpc.h>
+#include <thread>
+
+using namespace capnp;
+using namespace circt::esi::cosim;
+
+/// ------ EndpointServer definitions.
+
+EndpointServer::EndpointServer(Endpoint &ep) : endpoint(ep), open(true) {}
+EndpointServer::~EndpointServer() {
+  if (open)
+    endpoint.returnForUse();
+}
+
+/// This is the client polling for a message. If one is available, send it.
+/// TODO: implement a blocking call with a timeout.
+kj::Promise<void> EndpointServer::recv(RecvContext context) {
+  KJ_REQUIRE(open, "EndPoint closed already");
+  KJ_REQUIRE(!context.getParams().getBlock(),
+             "Blocking recv() not supported yet");
+
+  // Try to pop a message.
+  Endpoint::BlobPtr blob;
+  auto msgPresent = endpoint.getMessageToClient(blob);
+  context.getResults().setHasData(msgPresent);
+  if (msgPresent) {
+    KJ_REQUIRE(blob->size() % 8 == 0,
+               "Response msg was malformed. Size of response was not a "
+               "multiple of 8 bytes.");
+    // Copy the blob into a single segment.
+    auto segment =
+        kj::ArrayPtr<word>((word *)blob->data(), blob->size() / 8).asConst();
+    // Create a single-element array of segments.
+    auto segments = kj::heapArray({segment});
```

Comment:
What is kj::heapArray?

Reply:
It's the libkj heap allocated array. Actually, it's a helper function to a builder which returns a kj::Array<T>. I've replaced `auto` with the whole type to make it clearer.